Revise workflow guide: tesserocr, font processors, multistep processors (#189)
OCR-D/ocrd-website: site/en/workflows.md

<!-- END-EVAL -->

> ### Step 0.2: Font detection

Reviewer: According to https://dl.acm.org/doi/pdf/10.1145/3352631.3352640 @seuretm et al trained on a very large and diverse dataset, with both geometric and color-space random augmentation (contrast, brightness, hue, even compression artifacts and 2 different binarization methods), so IMHO it is safe to assume these models are quite robust themselves and better suited on the raw input (at least better than most external, general-purpose enhancement can deliver for the specific task).

But we should also mention that fact in the workflow guide, so people know the reason it is the first step.

> ### Step 0.2: Font detection

Author: > 
**Note:** `ocrd-typegroups-classifier` was trained on a very large and diverse
dataset, with both geometric and color-space random augmentation (contrast,
brightness, hue, even compression artifacts and 2 different binarization
methods), so it works best on the raw, *non-binarized* RGB image.

  * `other_font`
  * `not_a_font`

**Note:** `ocrd-typegroups-classifier` was trained on a very large and diverse
dataset, with both geometric and color-space random augmentation (contrast,
brightness, hue, even compression artifacts and 2 different binarization
methods), so it works best on the raw, *non-binarized* RGB image.

**Note:** `ocrd-typegroups-classifier` comes with a non-OCR-D CLI that allows

> for the generation of "heatmaps" on the page to visualize which regions of the page

Reviewer: Note: in the table below, you cannot use markdown hyperlinks, and this breaks table layout. (Cannot make suggestions out of context, sorry.)

### Step 7: Region segmentation

<!-- BEGIN-EVAL sed -n '0,/^## Notes/ p' ./repo/ocrd-website.wiki/Workflow-Guide-region-segmentation.md|sed '$d' -->
In this processing step, an (optimized) document image is taken as an input and the
image is segmented into the various regions, including columns.
Segments are also classified, either coarse (text, separator, image, table, ...) or fine-grained (paragraph, marginalia, heading, ...).

**Note:** If you use `ocrd-tesserocr-segment-region`, which uses only bounding
boxes instead of polygon coordinates, then you should post-process via
`ocrd-segment-repair` with `plausibilize=True` to obtain better results without

> large overlaps. Alternatively, consider using the all-in-one [`ocrd-tesserocr-recognize`](#step-x-multistep)
> processor that can do region segmentation, line segmentation and text recognition
> in one step and can query tesseract for the generally more precise polygon region outlines
> instead of more coarse bounding boxes.

Reviewer: ```suggestion
large overlaps. _Alternatively_, consider using the all-in-one capabilities of [`ocrd-tesserocr-segment` and `ocrd-tesserocr-recognize`](#step-x-multistep), which can do region segmentation and line segmentation (and optionally also text recognition) in one step by querying Tesseract's internal iterator (accessing the more precise polygon outlines instead of just coarse bounding boxes with lots of hard-to-recover overlap).
```

> **Note:** The `ocrd-tesserocr-recognize`, `ocrd-sbb-textline-detector` and

Reviewer: ```suggestion
**Note:** The `ocrd-tesserocr-segment`, `ocrd-tesserocr-recognize`, `ocrd-sbb-textline-detector` and
```

      <td>ocrd-tesserocr-segment-region -P overwrite_regions true -P find_tables false</td>
      <td>ocrd-tesserocr-recognize -P textequiv_level none -P segmentation_level region -P overwrite_segments true -P find_tables false</td>
    </tr>
    <tr>
      <td>ocrd-tesserocr-segment-table -P overwrite_cells true</td>
      <td>ocrd-tesserocr-recognize -P textequiv_level none -P segmentation_level cell -P overwrite_segments true</td>
    </tr>
    <tr>
      <td>ocrd-tesserocr-segment-line -P overwrite_lines true</td>
      <td>ocrd-tesserocr-recognize -P textequiv_level none -P segmentation_level line -P overwrite_segments true</td>
    </tr>
    <tr>
      <td>ocrd-tesserocr-segment-word -P overwrite_words true</td>
      <td>ocrd-tesserocr-recognize -P textequiv_level none -P segmentation_level word -P overwrite_segments true</td>
    </tr>
  </tbody>
</table>

**Note:** The three parameters `segmentation_level`, `textequiv_level` and
`model` define the behavior of `ocrd-tesserocr-segment`:

* `segmentation_level` determines the *highest level* to segment.
* `textequiv_level` determines the *lowest level* to segment. Use `"none"` to segment until the lowest level, i.e. glyphs.

> * `model` determines the model to use for text recognition.

Reviewer: ```suggestion
* `model` determines the model to use for text recognition. Use `""` or do not set at all to disable recognition, i.e. only analyse layout.
```

> To segment existing regions into lines (and only lines): `segmentation_level="line"`, `textequiv_level="line"`, `model=""`

Reviewer: ```suggestion
Examples:
- To segment existing regions into lines (and only lines) only: `segmentation_level="line"`, `textequiv_level="line"`, `model=""`
```

> 
> To segment existing regions into lines (and only lines) and recognize text: `segmentation_level="line"`, `textequiv_level="line"`, `model="Fraktur"`

Reviewer: ```suggestion
- To segment existing regions into lines (and only lines) and recognize text: `segmentation_level="line"`, `textequiv_level="line"`, `model="Fraktur"`

For detailed descriptions of behaviour and options, see [Readme file](https://github.com/OCR-D/ocrd_tesserocr/blob/master/README.md) and `ocrd-tesserocr-recognize/segment/segment-region/segment-table/segment-line/segment-word -h` help.
```

### Step 7: Region segmentation

<!-- BEGIN-EVAL sed -n '0,/^## Notes/ p' ./repo/ocrd-website.wiki/Workflow-Guide-region-segmentation.md|sed '$d' -->
In this processing step, an (optimized) document image is taken as an input and the
image is segmented into the various regions, including columns.
Segments are also classified, either coarse (text, separator, image, table, ...) or fine-grained (paragraph, marginalia, heading, ...).

> **Note:** If you use `ocrd-tesserocr-segment-region`, which uses only bounding

Reviewer: I agree, but this is all still quite fresh, so we should keep all options in there for the moment. The current ordering makes explanation/description easier, but requires a longer read. Practically, yes, I would recommend single-level segmentation with `shrink_polygons=True` over multi-level segmentation (because of workflow freedom) over singe-level segmentation with external post-processing.

> **Note:** If you use `ocrd-tesserocr-segment-region`, which uses only bounding

Author: Agreed that the change is quite recent (even unmerged in the case of shrink_polygons) but @EEngl52 has a point about users chosing the first thing they see and sticking with it.

> **Note:** If you use `ocrd-tesserocr-segment-region`, which uses only bounding

Reviewer: Sure, go ahead. Whatever formulation you come up with, I'll try to correct if necessary.

> **Note:** If you use `ocrd-tesserocr-segment-region`, which uses only bounding

Author: I stole part of your nice summary :) with `--author bertsky` though, so that makes it okay. 👼

> **Note:** If you use `ocrd-tesserocr-segment-region`, which uses only bounding

Reviewer: Cannot find it, sorry.

    </tr>
  </thead>
  <tbody>
    <tr data-processor="ocrd-tesserocr-segment">
      <td>ocrd-tesserocr-segment</td>

>       <td><code>-P find_tables false -P shrink_polygon true -P segmentation_level region -P textequiv_level none</code></td>

Reviewer: Note: only `recognize` exposes the `level` params. And since we have `shrink_polygons` now, I think we can safely recommend _both_ `ocrd-tesserocr-segment-region` and `ocrd-tesserocr-segment` now.
```suggestion
      <td><code>-P find_tables false -P shrink_polygons true</code></td>
```

      <td>Recommended. Will reuse internal tesseract iterators to produce a complete segmentation with tight polygons instead of bounding boxes where possible</td>

>       <td><code>ocrd-tesserocr-segment -I OCR-D-DEWARP-PAGE -O OCR-D-SEG</code></td>

Reviewer: ```suggestion
      <td><code>ocrd-tesserocr-segment -I OCR-D-DEWARP-PAGE -O OCR-D-SEG -P find_tables false -P shrink_polygons true</code></td>
```

      <td></td>
    <td><code>ocrd-cis-ocropy-segment -I OCR-D-DEWARP-PAGE -O OCR-D-SEG-LINE -P level-of-operation page</code></td>
    </tr>
    <tr data-processor="ocrd-tesserocr-segment-region">
      <td>ocrd-tesserocr-segment-region</td>
      <td><code>-P find_tables false</code></td>
      <td>Recommended</td>

>       <td><code>ocrd-tesserocr-segment-region -I OCR-D-DEWARP-PAGE -O OCR-D-SEG-REG</code></td>

Reviewer: ```suggestion
      <td><code>ocrd-tesserocr-segment-region -I OCR-D-DEWARP-PAGE -O OCR-D-SEG-REG -P find_tables false -P shrink_polygons true</code></td>
```

input file group, and a TextLine element with the resulting polygon
outline is added to the annotation of the output PAGE.

**Note:** If you use `ocrd-cis-ocropy-segment`, you can directly go on with [Step 13](#step-13-dewarping-on-line-level).

**Note:** If you use `ocrd-tesserocr-segment-line`, which uses only bounding boxes instead of polygon coordinates,

> then you should post-process with the processors described in [Step 12](#step-12-resegmentation-line-level). Alternatively,
> you can use the [`ocrd-tesserocr-recognize` all-in-one processor](#step-x-multistep) that can do line segmentation, region segmentation
> and recognition in one step.

Reviewer: ```suggestion
then you should post-process with the processors described in [Step 12](#step-12-resegmentation-line-level). 
_Alternatively_, consider using the all-in-one capabilities of [`ocrd-tesserocr-recognize`](#step-x-multistep), which can do line segmentation and text recognition in one step by querying Tesseract's internal iterator (accessing the more precise polygon outlines instead of just coarse bounding boxes with lots of hard-to-recover overlap).
```

input file group, and a TextLine element with the resulting polygon
outline is added to the annotation of the output PAGE.

**Note:** If you use `ocrd-cis-ocropy-segment`, you can directly go on with [Step 13](#step-13-dewarping-on-line-level).

**Note:** If you use `ocrd-tesserocr-segment-line`, which uses only bounding
boxes instead of polygon coordinates, then you should post-process with the
processors described in [Step 12](#step-12-resegmentation-line-level).
_Alternatively_, consider using the all-in-one capabilities of
[`ocrd-tesserocr-recognize`](#step-7-region-segmentation), which can do line segmentation
and text recognition in one step by querying Tesseract's internal iterator
(accessing the more precise polygon outlines instead of just coarse bounding

> boxes with lots of hard-to-recover overlap).

Reviewer: ```suggestion
boxes with lots of hard-to-recover overlap). _Alternatively_, run with `shrink_polygons=True` (accessing that same iterator to calculate convex hull polygons).
```

input file group, and a TextLine element with the resulting polygon
outline is added to the annotation of the output PAGE.

> **Note:** If you use `ocrd-cis-ocropy-segment`, you can directly go on with [Step 13](#step-13-dewarping-on-line-level).

Reviewer 2: this is already explained in the old text and could be left out

Reviewer: > this is already explained in the old text and could be left out

What do you mean _in the old text_? AFAICS that note is still valid, because there exists a `level-of-operation=region` mode for `ocrd-cis-ocropy-segment` which needs no post-processing (step 12 would be redundant).

Author: > What do you mean in the old text?

I think @EEngl52 was referring to the paragraph in region segmentation that tells users to skip region post-processing and line segmentation when using one of the region-and-line segmenters. But you're right, there's a valid reason why people want to skip step 12 if not required.

**Note:** If you use `ocrd-tesserocr-segment-line`, which uses only bounding
boxes instead of polygon coordinates, then you should post-process with the
processors described in [Step 12](#step-12-resegmentation-line-level).
_Alternatively_, consider using the all-in-one capabilities of
[`ocrd-tesserocr-recognize`](#step-7-region-segmentation), which can do line segmentation
and text recognition in one step by querying Tesseract's internal iterator
(accessing the more precise polygon outlines instead of just coarse bounding
boxes with lots of hard-to-recover overlap). _Alternatively_, run with `shrink_polygons=True` (accessing that same iterator to calculate convex hull polygons).

**Note:** As described in [Step 7](#step-7-page-segmentation), `ocrd-sbb-textline-detector` and `ocrd-cis-ocropy-segment` do not only segment

> the page, but also the text lines within the detected text regions in one step. Therefore with those (and only with those!) processors you don’t

Reviewer 2: ```suggestion
the page, but also the text lines within the detected text regions in one step. Therefore with those (and only with those!) processors you can directly go on with Step 13.
```

Reviewer: Leaving out step 11 itself (and step 12), because it does not apply anymore, is not the same as continuing directly with step 13. Also, your edit keeps the next line untouched (you can make multi-line suggestions by dragging the plus symbol downwards across a range of lines).

Reviewer 2: ah, sry, misread it

revision](https://github.com/tesseract-ocr/tessdata/commit/3cf1e2df1fe1d1da29295c9ef0983796c7958b7d)
before the [LSTM
models](https://github.com/tesseract-ocr/tessdata/commit/4592b8d453889181e01982d22328b5846765eaad)
replaced them, usually under the same name. (Thus, `deu.traineddata` used to be
a rule-based model but now is an LSTM model. `deu-frak.traineddata` is still
only available as rule-based model and was complemented by the new LSTM models
`frk.traineddata` and `script/Fraktur.traineddata`.) If you do need one of the
models that was replaced completely, then you should at least rename the old
one (e.g. to `deu3.traineddata`).

#### Available processors

<table class="processor-table">
  <thead>
    <tr>
      <th>Processor</th>
      <th>Parameter</th>
      <th>Remarks</th>
    <th>Call</th>
    </tr>
  </thead>
  <tbody>
    <tr data-processor="ocrd-tesserocr-fontshape">
      <td>ocrd-tesserocr-fontshape</td>
      <td><code>-P model osd -P padding 2</code></td>
      <td>Download other pre-LSTM models [from GitHub](https://github.com/tesseract-ocr/tessdata/commit/3cf1e2df1fe1d1da29295c9ef0983796c7958b7d)</td>

>       <td><code>ocrd-tesserocr-fontshape -I SEG-WORD -O SEG-WORD-FONT</code></td>

Reviewer: ```suggestion
      <td><code>ocrd-tesserocr-fontshape -I OCR-D-OCR -O OCR-D-OCR-FONT</code></td>
```

<!-- END-EVAL -->

### Step 14.1: Font style annotation

<!-- BEGIN-EVAL sed -n '0,/^## Notes/ p' ./repo/ocrd-website.wiki/Workflow-Guide-font-style-annotation.md|sed '$d' -->

> These processors can determine the font style (e.g. *italic*, **bold**,

Reviewer 2: ```suggestion
This processor can determine the font style (e.g. *italic*, **bold**,
```

<ins>underlined</ins>) and font family text recognition results.

> **Note:** `ocrd-tesserocr-fontshape` can either use existing segmentation or

Reviewer 2: ```suggestion
`ocrd-tesserocr-fontshape` can either use existing segmentation or
```

    </tr>
  </tbody>

<!-- END-EVAL -->

# Recommendations

<!-- BEGIN-INCLUDE ./repo/ocrd-website.wiki/Workflow-Guide-recommendations.md -->
In order to facilitate the usage of OCR-D and the configuration of workflows, we provide two workflows
which can be used as a start for your OCR-D-tests. They were determined by testing the processors listed
above on selected pages of some prints from the 17th and 18th century.

The results vary quite a lot from page to page. In most cases, segmentation is a problem.

Note that for our test pages, not all steps described above werde needed to obtain the best results.
Depending on your particular images, you might want to include those processors again for better results.

We are currently working on regression tests with the help of which we will be able to provide more profound
workflows soon, which will replace those interm solutions. 

## Minimal workflow

Since `ocrd-tesserocr-recognize` can do all line segmentation, region
segmentation, table detection and even binarization (Otsu), just like the
upstream `tesseract` command line tool, it's a good single-step workflow to get
a baseline result to compare to granular workflows.

> 

Reviewer 2: Be aware that you will most likely obtain significantly better results by configuring a more complex workflow like e.g. the workflows below.

Author: cc61fcd6

  </tbody>
</table>

### Step X: Multi-step

<!-- BEGIN-EVAL sed -n '0,/^## Notes/ p' ./repo/ocrd-website.wiki/Workflow-Guide-multistep.md|sed '$d' -->
The procecssors in this group do more than one step in an OCR-D workflow. This
allows these processors to optimize the data flow, e.g. reusing more precise
internal data structures, instead of relying on the serialization of results by
a previous step. The downside of these processors is that they are less
interoperable with the granular processors, removing a certain amount of
control from the user and making the implementations more complex.

**Note:** All the `ocrd-tesserocr-segment*` processors internally delegate to
`ocrd-tesserocr-recognize`, so you can replace calls to these task-specific
processors with calls to `ocrd-tesserocr-recognize` with specific parameters:

<table>
  <thead><tr><th>processor call</th><th><code>ocrd-tesserocr-recognize</code> parameters</th></tr></thead>
  <tbody>
    <tr>

>       <td>ocrd-tesserocr-segment-region -P overwrite_regions true -P crop_polygons</td>
>       <td>ocrd-tesserocr-recognize -P textequiv_level none -P segmentation_level region -P overwrite_segments true -P block_polygons</td>

Reviewer: `crop_polygons` is [unusable](https://github.com/OCR-D/ocrd_tesserocr/issues/98) at the moment, plus you would need to at least set a _value_ here. How about another param as example:

```suggestion
      <td>ocrd-tesserocr-segment-region -P overwrite_regions true -P find_tables false</td>
      <td>ocrd-tesserocr-recognize -P textequiv_level none -P segmentation_level region -P overwrite_segments true -P find_tables false</td>
```

    </tr>
    <tr>
      <td>ocrd-tesserocr-segment-table -P overwrite_cells true</td>
      <td>ocrd-tesserocr-recognize -P textequiv_level none -P segmentation_level cell -P overwrite_segments true</td>
    </tr>
    <tr>
      <td>ocrd-tesserocr-segment-line -P overwrite_lines true</td>
      <td>ocrd-tesserocr-recognize -P textequiv_level none -P segmentation_level line -P overwrite_segments true</td>
    </tr>
    <tr>
      <td>ocrd-tesserocr-segment-word -P overwrite_words true</td>
      <td>ocrd-tesserocr-recognize -P textequiv_level none -P segmentation_level word -P overwrite_segments true</td>
    </tr>
  </tbody>
</table>

<!-- END-EVAL -->

### Step X: Font annotation

<!-- BEGIN-EVAL sed -n '0,/^## Notes/ p' ./repo/ocrd-website.wiki/Workflow-Guide-font.md|sed '$d' -->
These processors can determine the font family (e.g. Antiqua, Fraktur,
Schwabacher) or font style (e.g. *italic*, **bold**).

**Note:** `ocrd-tesserocr-fontshape` can either use existing segmentation or segment on-demand. It can detect the following font styles:
  * `fontSize`
  * `fontFamily`
  * `bold`
  * `italic`
  * `underlined`
  * `monospace`
  * `serif`

> 

Reviewer: ```suggestion

**Note:** `ocrd-tesserocr-fontshape` needs the old, pre-LSTM models to work at all. You can use the pre-installed `osd` (which is purely rule-based), but there might be better alternatives for your language and script. You can still get the old models from Tesseract's Github repo at the [last revision](https://github.com/tesseract-ocr/tessdata/commit/3cf1e2df1fe1d1da29295c9ef0983796c7958b7d) before the [LSTM models](https://github.com/tesseract-ocr/tessdata/commit/4592b8d453889181e01982d22328b5846765eaad) replaced them, usually under the same name. (Thus, `deu.traineddata` used to be a rule-based model but now is an LSTM model. `deu-frak.traineddata` is still only available as rule-based model and was complemented by the new LSTM models `frk.traineddata` and `script/Fraktur.traineddata`.) If you do need one of the models that was replaced completely, then you should at least rename the old one (e.g. to `deu3.traineddata`).
```

**Note:** `ocrd-typegroups-classifier` can only annotate font families on page level but can detect a wider variety of fonts, including the confidence value (separated by colon). Supported `fontFamily` values:
  * `Antiqua`
  * `Bastarda`
  * `Fraktur`
  * `Gotico`-Antiqua
  * `Greek`
  * `Hebrew`
  * `Italic`
  * `Rotunda`
  * `Schwabacher`
  * `Textura`
  * `other_font`
  * `not_a_font`

> **Note:** `ocrd-typegroups-classifier` only works on *non-binarized* RGB images.

Reviewer: Is that something the processor implementation can take care of (silently)?

Author: How do you mean "take care silently"? We could add defensive strategy to just bail out if provided bitonal images, but that would be the opposite of silent.

Reviewer: At the moment, `ocrd-typegroups-classifier` does not use core's image API:

https://github.com/seuretm/ocrd_typegroups_classifier/blob/a5a205bb2f9fce533ed195c76f5f23f648acf48b/ocrd_typegroups_classifier/processor.py#L38

I have already created a [PR](https://github.com/seuretm/ocrd_typegroups_classifier/pull/16) for this long ago but received no response.

Now, when using the image API, it's of course always possible to get the raw input image, simply by adding `feature_filter="binarized"` (which is always guaranteed to yield a result).

Author: There has been a bit of backlog with the PR, @seuretm doesn't have time for it apparently. Since I've been making PyPI releases from unmerged PR of it, maybe we should continue development in a fork until @seuretm (hopefully) returns to maintaining it.

#### Available processors

<table class="processor-table">
  <thead>
    <tr>
      <th>Processor</th>
      <th>Parameter</th>
      <th>Remarks</th>
    <th>Call</th>
    </tr>
  </thead>
  <tbody>
    <tr data-processor="ocrd-tesserocr-fontshape">
      <td>ocrd-tesserocr-fontshape</td>
      <td><code>-P model osd -P padding 2</code></td>

>       <td></td>

Reviewer: ```suggestion
      <td>Download other pre-LSTM models [from GitHub](https://github.com/tesseract-ocr/tessdata/commit/3cf1e2df1fe1d1da29295c9ef0983796c7958b7d)</td>
```

      <td>ocrd-typegroups-classifier</td>
      <td><code>-P network /path/to/densenet121.tgc</code></td>
      <td>Download [`densenet121.tgc` from GitHub](https://github.com/seuretm/ocrd_typegroups_classifier/raw/master/ocrd_typegroups_classifier/models/densenet121.tgc)</td>
      <td><code>ocrd-typegroups-classifier -I OCR-D-IMG -O OCR-D-IMG-FONTS</code></td>
    </tr>
  </tbody>

<!-- END-EVAL -->

# Recommendations

<!-- BEGIN-INCLUDE ./repo/ocrd-website.wiki/Workflow-Guide-recommendations.md -->
In order to facilitate the usage of OCR-D and the configuration of workflows, we provide two workflows
which can be used as a start for your OCR-D-tests. They were determined by testing the processors listed
above on selected pages of some prints from the 17th and 18th century.

The results vary quite a lot from page to page. In most cases, segmentation is a problem.

Note that for our test pages, not all steps described above werde needed to obtain the best results.
Depending on your particular images, you might want to include those processors again for better results.

We are currently working on regression tests with the help of which we will be able to provide more profound

> workflows soon, which will replace those interm solutions. 

Reviewer: ```suggestion
workflows soon, which will replace those interim solutions. 
```

## Minimal workflow

> Since `ocrd-tesserocr-recognize` can do all line segmentation, region
> segmentation, table detection and even binarization (Otsu), just like the

Reviewer: ```suggestion
Since `ocrd-tesserocr-recognize` can do all binarization (Otsu), region
segmentation, table recognition, line segmentation and text recognition at once, just like the
```

upstream `tesseract` command line tool, it's a good single-step workflow to get
a baseline result to compare to granular workflows.

**Note:** Be aware that you will most likely obtain significantly better
results by configuring a more granular workflow like e.g. the
[workflows](#best-results-for-selected-pages)
[below](#good-results-for-slower-processors).

<table class="processor-table">
  <thead>
    <tr>
      <th>Step</th>
      <th>Processor</th>
      <th>Parameter</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>1</td>
      <td>ocrd-tesserocr-recognize</td>

>       <td>-P segmentation_level line -P textequiv_level word -P find_tables true -P overwrite_segments true -P model GT4HistOCR_50000000.997_191951</td>

Reviewer: Note: `segmentation_level` is the **entry point** to segmentation; if you had no regions already, then `line` is too low and nothing will happen. I would disable and hide `overwrite_segments` here (we don't expect any existing segments).

```suggestion
      <td>-P segmentation_level region -P textequiv_level word -P find_tables true -P model GT4HistOCR_50000000.997_191951</td>
```

      <td>ocrd-typegroups-classifier</td>
      <td><code>-P network /path/to/densenet121.tgc</code></td>
      <td>Download [`densenet121.tgc` from GitHub](https://github.com/seuretm/ocrd_typegroups_classifier/raw/master/ocrd_typegroups_classifier/models/densenet121.tgc)</td>

>       <td><code>ocrd-typegroups-classifier -I MAX -O MAX-FONTS</code></td>

Reviewer 2: for consistency reasons I would change `MAX` to `OCR-D-IMG`

Depending on your particular images, you might want to include those processors again for better results.

We are currently working on regression tests with the help of which we will be able to provide more profound
workflows soon, which will replace those interim solutions. 

## Minimal workflow

Since `ocrd-tesserocr-recognize` can do binarization (Otsu), region
segmentation, table recognition, line segmentation and text recognition at once, just like the
upstream `tesseract` command line tool, it's a good single-step workflow to get
a baseline result to compare to granular workflows.

**Note:** Be aware that you will most likely obtain significantly better
results by configuring a more granular workflow like e.g. the
[workflows](#best-results-for-selected-pages)
[below](#good-results-for-slower-processors).

<table class="processor-table">
  <thead>
    <tr>
      <th>Step</th>
      <th>Processor</th>
      <th>Parameter</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>1</td>
      <td>ocrd-tesserocr-recognize</td>
      <td>-P segmentation_level region -P textequiv_level word -P find_tables true -P model GT4HistOCR_50000000.997_191951</td>
    </tr>
  </tbody>

> </table>

Reviewer 2: is there a call I can simply copy and paste somewhere?

Reviewer: You mean an _Example with ocrd-process_? Just copy the content of the table:

```suggestion
</table>

### Example with ocrd-process

    ocrd process "tesserocr-recognize -P segmentation_level region -P textequiv_level word -P find_tables true -P model GT4HistOCR_50000000.997_191951"

```

(Using space-indented command code syntax here, because code block inside suggestion block doesn't work; fingers crossed.)

Reviewer 2: perfect :+1:

Reviewer: > perfect

Did you want to commit the suggestion?

  "tesserocr-deskew -I OCR-D-BIN-DENOISE -O OCR-D-BIN-DENOISE-DESKEW -P operation_level page" \
  "cis-ocropy-segment -I OCR-D-BIN-DENOISE-DESKEW -O OCR-D-SEG-REG -P level-of-operation page" \
  "tesserocr-deskew -I OCR-D-SEG-REG -O OCR-D-SEG-REG-DESKEW" \

>   "cis-ocropy-dewarp -I OCR-D-SEG-REG-DESKEW -O OCR-D-SEG-LINE-RESEG-DEWARP" \

Reviewer: This change does not work quite like that.

Why deskew on the region level after already segmenting lines? If you don't also segment lines again afterwards, the regions' orientation will be ignored. 

If you expect regions with 90° or 270°, then you need keep the ocrd-cis-ocropy-segment step in there.

So either just remove the clipping steps, or also remove the region-deskewing step.

Author: I am not even sure where this change comes from, wasn't a deliberate choice to change the existing recommendations (though we should). But thanks for the notes, I'll adapt the workflow.

Author: removed clipping and deskewing from the "best" workflow, the second binarization from the "fast" workflow.

>       <td>ocrd-tesserocr-deskew</td>

Reviewer: Gone in the command list below, but still here! (Cf. [my comment last time](https://github.com/OCR-D/ocrd-website/pull/189#discussion_r536154337) …)

You need to remove region-level deskewing here, too, and renumber from 9 upwards.

>       <td>ocrd-tesserocr-segment-line</td>

Reviewer: This second (now third) workflow recommendation should really be changed now. With the all-in-one segmentation for Tesseract, we don't need repair, region-level deskewing and clipping anymore. Just do `ocrd-tesserocr-segment` and then dewarping. Also, when you advertise this for slower hardware, skip the overly slow, redundant skimage-binarize.

>     </tr>

Reviewer: Step 3 skimage-binarize is still in here (and also in the command list below). This is _slow_ and not strictly necessary. Just a reminder in case you did forget (cf. [my comment last time](https://github.com/OCR-D/ocrd-website/pull/189#discussion_r536159259)).

    <tr>
      <td>7</td>
      <td>ocrd-tesserocr-segment</td>

>       <td></td>

Reviewer: Also helpful here, because you have dewarping afterwards, so better polygonalize the textline coords.

```suggestion
      <td>-P shrink_polygons true</td>
```

    </tr>
    <tr>
    <tr>

>       <td>13</td>

Reviewer: Renumber?

    <tr>

>       <td>14</td>

Reviewer: Renumber?

Author: The idea is to reference the step number from the guide before the recommendation. I'll make internal links of ### ### the numbers to make that clearer.

  "skimage-binarize -I OCR-D-CROP -O OCR-D-BIN2 -P method li" \
  "skimage-denoise -I OCR-D-BIN2 -O OCR-D-BIN-DENOISE -P level-of-operation page" \
  "tesserocr-deskew -I OCR-D-BIN-DENOISE -O OCR-D-BIN-DENOISE-DESKEW -P operation_level page" \

>   "tesserocr-segment -I OCR-D-BIN-DENOISE-DESKEW -O OCR-D-SEG" \

Reviewer: ```suggestion
  "tesserocr-segment -I OCR-D-BIN-DENOISE-DESKEW -O OCR-D-SEG -P shrink_polygons true" \
```

  "cis-ocropy-dewarp -I OCR-D-SEG -O OCR-D-SEG-DEWARP" \

>   "tesserocr-recognize -I OCR-D-SEG-DEWARP -O OCR-D-OCR -P textequiv_level glyph -P overwrite_segments true -P model GT4HistOCR_50000000.997_191951}"

Reviewer: ```suggestion
  "tesserocr-recognize -I OCR-D-SEG-DEWARP -O OCR-D-OCR -P textequiv_level glyph -P overwrite_segments true -P model GT4HistOCR_50000000.997_191951"
```
